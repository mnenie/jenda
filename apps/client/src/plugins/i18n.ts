import { createI18n } from 'vue-i18n'
import { enZod, ruZod, zhZod } from '@/shared/libs/vee-validate/rules'
import { customPluralRule } from '@/shared/libs/i18n/plurals'
import { getCurrentLocale } from '@/shared/libs/i18n/utils/getLocale'

interface MessageSchema {
  [key: string]: any
}

const ru_ru = import.meta.glob(
  '@/shared/libs/i18n/locales/**/ru-RU.ts',
  { eager: true },
)
const en_us = import.meta.glob(
  '@/shared/libs/i18n/locales/**/en-US.ts',
  { eager: true },
)
const zh_cn = import.meta.glob(
  '@/shared/libs/i18n/locales/**/zh-CN.ts',
  { eager: true },
)

function loadLang(modules: Record<string, any>) {
  const messages: { [key: string]: string } = {}

  Object.keys(modules).forEach((module) => {
    Object.assign(messages, { ...modules[module].default })
  })
  return messages
}

const messages = {
  'en-US': {
    ...loadLang(en_us),
    errors: {
      ...enZod,
    },
  },
  'ru-RU': {
    ...loadLang(ru_ru),
    errors: {
      ...ruZod,
    },
  },
  'zh-CN': {
    ...loadLang(zh_cn),
    errors: {
      ...zhZod,
    },
  },
}

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  pluralRules: {
    'ru-RU': {
      // @ts-expect-error i18n shema
      customPluralRule,
    },
  },
  messages,
})

export default i18n
