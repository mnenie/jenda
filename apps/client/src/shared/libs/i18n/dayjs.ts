import { watch } from 'vue'
import dayjs from 'dayjs'
import 'dayjs/locale/en'
import 'dayjs/locale/ru'
import 'dayjs/locale/zh-cn'
import type { I18n } from 'vue-i18n'

export default function useDayJsLocalize<
  T extends I18n<Record<string, string>, Record<string, string>>,
>(
  i18n: T,
) {
  const dayJsLocaleMap: Record<string, string> = {
    'en-US': 'en',
    'ru-RU': 'ru',
    'zh-CN': 'zh-cn',
  }

  watch(
    // @ts-ignore
    () => i18n.global!.locale.value,
    (newLocale) => {
      dayjs.locale(dayJsLocaleMap[newLocale as keyof typeof dayJsLocaleMap])
    },
    { immediate: true },
  )
}
