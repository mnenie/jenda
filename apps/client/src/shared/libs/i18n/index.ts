import { createI18n, type I18n } from 'vue-i18n'
import { enZod, ruZod, zhZod } from '../vee-validate/rules'
import { enLocale, ruLocale, zhLocale } from './locales'
import { customPluralRule } from './plurals'
import useDayJsLocalize from './dayjs'
import { getCurrentLocale } from './utils/getLocale'

export type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
    errors: {
      ...enZod,
    },
  },
  'ru-RU': {
    ...ruLocale,
    errors: {
      ...ruZod,
    },
  },
  'zh-CN': {
    ...zhLocale,
    errors: {
      ...zhZod,
    },
  },
}

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  pluralRules: {
    'ru-RU': {
      // @ts-expect-error i18n shema
      customPluralRule,
    },
  },
  messages,
})

useDayJsLocalize(i18n as I18n)

export default i18n
