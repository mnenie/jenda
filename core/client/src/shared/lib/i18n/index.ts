import { useCookies } from '@vueuse/integrations/useCookies'
import { computed } from 'vue'
import { createI18n } from 'vue-i18n'
import enLocale from './locales/en-US'
import ruLocale from './locales/ru-RU'
import zhLocale from './locales/zh-CN'

type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
  },
  'ru-RU': {
    ...ruLocale,
  },
  'zh-CN': {
    ...zhLocale,
  },
}
const cookies = useCookies()

const getCurrentLocale = computed(() => {
  const cookieLanguage = cookies.get('i18n')
  if (cookieLanguage) {
    return cookieLanguage
  }
  if (navigator.language.split('-')[0] === 'ru') {
    return 'ru-RU'
  }
  return 'en-US'
})

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  messages,
})

export default i18n
