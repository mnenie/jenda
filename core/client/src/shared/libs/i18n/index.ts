import { computed } from 'vue'
import { createI18n } from 'vue-i18n'

import en from '../vee-validate/rules/en'
import ru from '../vee-validate/rules/ru'
import zh from '../vee-validate/rules/zh'
import enLocale from './locales/en-US'
import ruLocale from './locales/ru-RU'
import zhLocale from './locales/zh-CN'

export type MessageSchema = typeof enLocale

const messages = {
  'en-US': {
    ...enLocale,
    errors: {
      ...en,
    },
  },
  'ru-RU': {
    ...ruLocale,
    errors: {
      ...ru,
    },
  },
  'zh-CN': {
    ...zhLocale,
    errors: {
      ...zh,
    },
  },
}

const getCurrentLocale = computed(() => {
  const storageLanguage = localStorage.getItem('i18n')
  if (storageLanguage) {
    return storageLanguage
  }
  if (navigator.language.split('-')[0] === 'ru') {
    return 'ru-RU'
  }
  return 'en-US'
})

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale.value,
  globalInjection: true,
  messages,
})

export default i18n
