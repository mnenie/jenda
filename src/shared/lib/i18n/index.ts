import { createI18n } from 'vue-i18n';
import { useCookies } from '@vueuse/integrations/useCookies';
import enLocale from './locales/en-US';
import ruLocale from './locales/ru-RU';
import zhLocale from './locales/zh-CN';

type MessageSchema = typeof enLocale;

const messages = {
  'en-US': {
    ...enLocale
  },
  'ru-RU': {
    ...ruLocale
  },
  'zh-CN': {
    ...zhLocale
  }
};
const cookies = useCookies();

const getCurrentLocale = () => {
  const cookieLanguage = cookies.get('i18n');
  if (cookieLanguage) {
    return cookieLanguage;
  }

  return 'en-US';
};

const i18n = createI18n<[MessageSchema], 'en-US' | 'ru-RU' | 'zh-CN'>({
  legacy: false,
  locale: getCurrentLocale(),
  messages: messages
});

export default i18n;
