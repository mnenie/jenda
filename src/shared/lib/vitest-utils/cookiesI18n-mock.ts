import { vi } from 'vitest';

vi.mock('@vueuse/integrations/useCookies', () => {
  return {
    useCookies: () => ({
      get(key: string) {
        return key === 'i18n' ? 'en-US' : undefined;
      }
    })
  };
});
