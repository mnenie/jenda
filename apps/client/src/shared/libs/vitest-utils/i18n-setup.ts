import { config } from '@vue/test-utils'
import i18n from '@/plugins/i18n'

config.global.plugins = [...(config.global.plugins || []), i18n]
