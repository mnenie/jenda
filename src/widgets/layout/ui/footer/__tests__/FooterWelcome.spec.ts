import { describe, it, expect } from 'vitest';
import { shallowMount } from '@vue/test-utils';
import '@/shared/lib/vitest-utils/cookiesI18n-mock';
import i18n from '@/shared/lib/i18n';
import FooterWelcome from '../AppFooter.vue';

describe('tests for FooterWelcome.vue', () => {
  const wrapper = shallowMount(FooterWelcome, {
    global: {
      plugins: [i18n],

      mocks: {
        t: (key: string) => {
          const translations: Record<string, string> = {
            'welcome.footer': 'footer'
          };
          return translations[key];
        }
      }
    }
  });

  // snapshot is covered needs for i18n I think ;)
  it('should be render correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
