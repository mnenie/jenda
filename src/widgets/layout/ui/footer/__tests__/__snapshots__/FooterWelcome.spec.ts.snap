// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests for FooterWelcome.vue > should be render correctly 1`] = `
"<footer class="footer">
  <div class="container">
    <p class="text-sm"><span class="brand">Jenda</span> - cloud-based program for task management. The source code is available on <a class="link"> Github </a></p>
  </div>
</footer>"
`;
